File: app/src/main/java/com/zaknong/airus/ui/MainActivity.java
package com.zaknong.airus.ui;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.view.View;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;
import androidx.navigation.ui.NavigationUI;

import com.bumptech.glide.Glide;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.zaknong.airus.R;
import com.zaknong.airus.database.entity.Song;
import com.zaknong.airus.permissions.PermissionManager;
import com.zaknong.airus.scanner.MediaScanner;
import com.zaknong.airus.scanner.TagEnricher;
import com.zaknong.airus.service.PlayerService;
import com.zaknong.airus.service.PlayerState;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MainActivity";

    // =========================================================
    // Views
    // =========================================================
    private BottomNavigationView bottomNav;
    private View                 nowPlayingBar;
    private ImageView            miniAlbumArt;
    private TextView             miniTitle;
    private TextView             miniArtist;
    private ImageButton          miniPlayPause;
    private ImageButton          miniNext;
    private ProgressBar          miniProgress;

    // =========================================================
    // Service binding
    // =========================================================
    private PlayerService playerService;
    private boolean       serviceBound = false;

    private final ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder binder) {
            PlayerService.LocalBinder localBinder =
                    (PlayerService.LocalBinder) binder;
            playerService = localBinder.getService();
            serviceBound  = true;
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            serviceBound = false;
        }
    };

    // =========================================================
    // Permission + Scanner
    // =========================================================
    private PermissionManager permissionManager;

    // =========================================================
    // Lifecycle
    // =========================================================

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        getWindow().getDecorView().setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE |
                        View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);

        setContentView(R.layout.activity_main);

        // ← Start service dulu sebelum apapun
        startPlayerService();

        permissionManager = new PermissionManager(this, granted -> {
            if (granted) {
                startMediaScan();
            } else {
                showPermissionEmptyState();
            }
        });

        bindViews();
        setupNavigation();
        observePlayerState();

        permissionManager.requestStoragePermission();
    }

    @Override
    protected void onStart() {
        super.onStart();
        // Bind ke PlayerService
        Intent intent = new Intent(this, PlayerService.class);
        bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE);
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (serviceBound) {
            unbindService(serviceConnection);
            serviceBound = false;
        }
    }

    // =========================================================
    // Setup
    // =========================================================

    private void bindViews() {
        bottomNav    = findViewById(R.id.bottom_nav);
        nowPlayingBar = findViewById(R.id.now_playing_bar);
        miniAlbumArt  = nowPlayingBar.findViewById(R.id.mini_album_art);
        miniTitle     = nowPlayingBar.findViewById(R.id.mini_title);
        miniArtist    = nowPlayingBar.findViewById(R.id.mini_artist);
        miniPlayPause = nowPlayingBar.findViewById(R.id.mini_play_pause);
        miniNext      = nowPlayingBar.findViewById(R.id.mini_next);
        miniProgress  = nowPlayingBar.findViewById(R.id.mini_progress);

        // Sembunyikan NowPlayingBar sampai ada lagu
        nowPlayingBar.setVisibility(View.GONE);
    }

    private void setupNavigation() {
        NavHostFragment navHostFragment = (NavHostFragment)
                getSupportFragmentManager()
                        .findFragmentById(R.id.nav_host_fragment);

        NavController navController = navHostFragment.getNavController();
        NavigationUI.setupWithNavController(bottomNav, navController);
    }

    // =========================================================
    // Observe PlayerState — update mini player
    // =========================================================

    private void observePlayerState() {
        PlayerState state = PlayerState.getInstance();

        // Update mini player saat lagu berubah
        state.getCurrentSong().observe(this, song -> {
            if (song == null) {
                nowPlayingBar.setVisibility(View.GONE);
                return;
            }
            nowPlayingBar.setVisibility(View.VISIBLE);
            updateMiniPlayer(song);
        });

        // Update icon play/pause
        state.getPlaybackState().observe(this, playbackState -> {
            boolean isPlaying = playbackState == PlayerState.State.PLAYING;
            miniPlayPause.setImageResource(
                    isPlaying ? R.drawable.ic_pause : R.drawable.ic_play);
        });

        // Update progress bar
        state.getPosition().observe(this, posMs -> {
            Long durMs = state.getDuration().getValue();
            if (durMs != null && durMs > 0) {
                int progress = (int) (posMs * 1000L / durMs);
                miniProgress.setProgress(progress);
            }
        });
    }

    private void updateMiniPlayer(Song song) {
        miniTitle.setText(song.title);

        // Artis + format label di baris kedua
        String subtitle = song.artist;
        String fmt = song.getFormatLabel();
        if (fmt != null && !fmt.isEmpty()) {
            subtitle = subtitle + " · " + fmt;
        }
        miniArtist.setText(subtitle);

        // Album art via Glide
        if (song.albumArtPath != null) {
            Glide.with(this)
                    .load(song.albumArtPath)
                    .placeholder(R.drawable.ic_default_album_art)
                    .centerCrop()
                    .into(miniAlbumArt);
        } else {
            miniAlbumArt.setImageResource(R.drawable.ic_default_album_art);
        }

        // Tap bar → buka NowPlayingFragment
        nowPlayingBar.setOnClickListener(v -> openNowPlaying());

        // Tap play/pause
        miniPlayPause.setOnClickListener(v -> {
            if (serviceBound) {
                PlayerState.State current =
                        PlayerState.getInstance().getPlaybackState().getValue();
                if (current == PlayerState.State.PLAYING) {
                    playerService.pause();
                } else {
                    playerService.play();
                }
            }
        });

        // Tap next
        miniNext.setOnClickListener(v -> {
            if (serviceBound) playerService.skipToNext();
        });
    }

    // =========================================================
    // NowPlaying
    // =========================================================

    private void openNowPlaying() {
        NowPlayingFragment fragment = new NowPlayingFragment();
        fragment.show(getSupportFragmentManager(), "now_playing");
    }

    // =========================================================
    // Service + Scanner
    // =========================================================

    private void startPlayerService() {
        Intent intent = new Intent(this, PlayerService.class);
        startForegroundService(intent);
    }

    private void startMediaScan() {
        MediaScanner scanner = new MediaScanner(this);
        TagEnricher  enricher = new TagEnricher(this);

        scanner.getScanProgress().observe(this, progress -> {
            if (progress.isFinished) {
                // Scan selesai — mulai enrich metadata detail di background
                enricher.enrichAll();
            }
        });

        scanner.startScan();
    }

    private void showPermissionEmptyState() {
        // Akan kita tambahkan empty state UI di LibraryFragment
    }
}

File: app/src/main/java/com/zaknong/airus/ui/MainActivity.kt
package com.zaknong.airus.ui

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.pager.HorizontalPager
import androidx.compose.foundation.pager.rememberPagerState
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.text.font.FontWeight
import androidx.activity.compose.BackHandler
import androidx.compose.animation.core.animateDpAsState
import androidx.preference.PreferenceManager
import androidx.navigation.compose.NavHost
import androidx.navigation.compose.composable
import androidx.navigation.compose.currentBackStackEntryAsState
import androidx.navigation.compose.rememberNavController
import com.zaknong.airus.permissions.PermissionManager
import com.zaknong.airus.scanner.MediaScanner
import com.zaknong.airus.scanner.TagEnricher
import com.zaknong.airus.service.PlayerService
import com.zaknong.airus.service.PlayerState
import com.zaknong.airus.ui.component.COLLAPSED_ANCHOR
import com.zaknong.airus.ui.component.DISMISSED_ANCHOR
import com.zaknong.airus.ui.component.rememberBottomSheetState
import com.zaknong.airus.ui.theme.AirusTheme
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {

    private var playerService by mutableStateOf<PlayerService?>(null)
    private var serviceBound by mutableStateOf(false)

    private val serviceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, binder: IBinder?) {
            val localBinder = binder as PlayerService.LocalBinder
            playerService = localBinder.getService()
            serviceBound = true
        }

        override fun onServiceDisconnected(name: ComponentName?) {
            serviceBound = false
            playerService = null
        }
    }

    private lateinit var permissionManager: PermissionManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()

        startPlayerService()

        permissionManager = PermissionManager(this) { granted ->
            if (granted) {
                // startMediaScan() // Remove auto-scan on permission grant in onCreate
            }
        }

        setContent {
            AirusTheme {
                MainScreenContent(
                    playerService = playerService,
                    onShowEq = {
                        val eqFragment = EqFragment()
                        eqFragment.show(supportFragmentManager, "eq")
                    }
                )
            }
        }

        // permissionManager.requestStoragePermission() // Remove auto-request in onCreate
    }

    override fun onStart() {
        super.onStart()
        val intent = Intent(this, PlayerService::class.java)
        bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE)
    }

    override fun onStop() {
        super.onStop()
        if (serviceBound) {
            unbindService(serviceConnection)
            serviceBound = false
        }
    }

    private fun startPlayerService() {
        val intent = Intent(this, PlayerService::class.java)
        startForegroundService(intent)
    }

    private fun startMediaScan() {
        val scanner = MediaScanner(this)
        val enricher = TagEnricher(this)

        scanner.scanProgress.observe(this) { progress ->
            if (progress.isFinished) {
                enricher.enrichAll()
            }
        }
        scanner.startScan()
    }
}

@Composable
fun MainScreenContent(
    playerService: PlayerService?,
    onShowEq: () -> Unit
) {
    val navController = rememberNavController()
    val navBackStackEntry by navController.currentBackStackEntryAsState()
    val currentRoute = navBackStackEntry?.destination?.route

    val playerState = PlayerState.getInstance()
    val currentSong by playerState.currentSong.observeAsState()
    val playbackState by playerState.playbackState.observeAsState(PlayerState.State.IDLE)
    val position by playerState.position.observeAsState(0L)
    val duration by playerState.duration.observeAsState(0L)

    val context = androidx.compose.ui.platform.LocalContext.current
    val prefs = remember { PreferenceManager.getDefaultSharedPreferences(context) }
    val isFirstLaunch = remember { mutableStateOf(prefs.getBoolean("first_launch", true)) }

    val scanManager = remember { com.zaknong.airus.scanner.ScanManager.getInstance(context) }
    val isScanning by scanManager.getIsScanning().observeAsState(false)
    val scanProgressText by scanManager.getScanProgressText().observeAsState("")

    val pagerState = rememberPagerState(pageCount = { Screens.MainScreens.size })
    val scope = rememberCoroutineScope()
    var libraryCategory by remember { mutableStateOf<String?>(null) }
    var initialPlaylist by remember { mutableStateOf<com.zaknong.airus.database.entity.Playlist?>(null) }

    val isMainScreen = currentRoute == "main" || currentRoute == null

    // First Launch Logic
    LaunchedEffect(isFirstLaunch.value) {
        if (isFirstLaunch.value) {
            navController.navigate("scan_folders")
            prefs.edit().putBoolean("first_launch", false).apply()
            isFirstLaunch.value = false
        }
    }

    // Back Navigation Logic
    BackHandler(enabled = true) {
        if (libraryCategory != null) {
            libraryCategory = null
        } else if (pagerState.currentPage != 0) {
            scope.launch { pagerState.animateScrollToPage(0) }
        } else {
            (context as? android.app.Activity)?.finish()
        }
    }

    BoxWithConstraints(modifier = Modifier.fillMaxSize()) {
        val screenHeight = maxHeight
        val bottomSheetState = rememberBottomSheetState(
            dismissedBound = 0.dp,
            collapsedBound = 80.dp,
            expandedBound = screenHeight,
            initialAnchor = if (currentSong != null) COLLAPSED_ANCHOR else DISMISSED_ANCHOR
        )

        LaunchedEffect(currentSong) {
            if (currentSong != null && bottomSheetState.isDismissed) {
                bottomSheetState.collapse()
            }
        }

        Scaffold(
            bottomBar = {
                if (isMainScreen) {
                    NavigationBar(
                        containerColor = Color.Black,
                        contentColor = Color.White,
                        tonalElevation = 0.dp
                    ) {
                        Screens.MainScreens.forEachIndexed { index, screen ->
                            NavigationBarItem(
                                icon = {
                                    Icon(
                                        painter = painterResource(if (pagerState.currentPage == index) screen.iconIdActive else screen.iconIdInactive),
                                        contentDescription = stringResource(screen.titleId),
                                        modifier = Modifier.size(26.dp)
                                    )
                                },
                                label = null,
                                selected = pagerState.currentPage == index,
                                onClick = {
                                    scope.launch { pagerState.animateScrollToPage(index) }
                                },
                                colors = NavigationBarItemDefaults.colors(
                                    selectedIconColor = MaterialTheme.colorScheme.primary,
                                    unselectedIconColor = Color.Gray,
                                    indicatorColor = Color.Transparent
                                )
                            )
                        }
                    }
                }
            },
            containerColor = Color.Black
        ) { padding ->
            Box(modifier = Modifier.padding(padding)) {
                NavHost(
                    navController = navController,
                    startDestination = "main",
                    modifier = Modifier.fillMaxSize()
                ) {
                    composable("main") {
                        HorizontalPager(
                            state = pagerState,
                            modifier = Modifier.fillMaxSize(),
                            beyondViewportPageCount = 2
                        ) { page ->
                            when (Screens.MainScreens[page].route) {
                                "home" -> HomeScreen(
                                    onSongClick = { list, index ->
                                        playerService?.playQueue(list, index)
                                    },
                                    onPlaylistClick = { playlist ->
                                        initialPlaylist = playlist
                                        scope.launch { pagerState.animateScrollToPage(2) }
                                    },
                                    onMoreClick = { category ->
                                        libraryCategory = category
                                        initialPlaylist = null
                                        scope.launch { pagerState.animateScrollToPage(2) }
                                    }
                                )
                                "search" -> SearchScreen(onSongClick = { list, index ->
                                    playerService?.playQueue(list, index)
                                })
                                "library" -> LibraryScreen(
                                    onSongClick = { list, index ->
                                        playerService?.playQueue(list, index)
                                    },
                                    onNavigateToFolders = { navController.navigate("scan_folders") },
                                    initialCategory = libraryCategory,
                                    initialPlaylist = initialPlaylist
                                )
                            }
                        }
                    }
                    composable("scan_folders") { ScanFoldersScreen(onBack = { navController.popBackStack() }) }
                }
            }
        }

        if (isMainScreen) {
            Box(
                modifier = Modifier.fillMaxSize(),
                contentAlignment = Alignment.BottomCenter
            ) {
                // Background overlay for expanded state
                if (bottomSheetState.progress > 0.01f) {
                    Box(
                        modifier = Modifier
                            .fillMaxSize()
                            .background(Color.Black.copy(alpha = 0.6f * bottomSheetState.progress))
                            .clickable(enabled = bottomSheetState.progress > 0.9f) {}
                    )
                }

                val offsetY = 80.dp * (1f - bottomSheetState.progress)

                BottomSheetPlayer(
                    state = bottomSheetState,
                    currentSong = currentSong,
                    playbackState = playbackState,
                    position = position,
                    duration = duration,
                    onPlayPause = { if (playbackState == PlayerState.State.PLAYING) playerService?.pause() else playerService?.play() },
                    onNext = { playerService?.skipToNext() },
                    onPrevious = { playerService?.skipToPrevious() },
                    onSeek = { playerService?.seekTo(it) },
                    onToggleShuffle = { playerService?.toggleShuffle() },
                    onToggleRepeat = { playerService?.cycleRepeatMode() },
                    onToggleFavorite = { playerService?.toggleFavorite() },
                    onShowEq = onShowEq,
                    onSkipToQueueItem = { playerService?.skipToQueueItem(it) },
                    modifier = Modifier
                        .align(Alignment.BottomCenter)
                        .offset(y = -offsetY)
                )
            }
        }

        if (isScanning) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .background(Color.Black.copy(alpha = 0.8f))
                    .clickable(enabled = false) {}, // Blokir interaksi
                contentAlignment = Alignment.Center
            ) {
                Card(
                    colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceVariant),
                    shape = RoundedCornerShape(24.dp),
                    modifier = Modifier.padding(32.dp)
                ) {
                    Column(
                        modifier = Modifier.padding(24.dp),
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        CircularProgressIndicator(
                            color = MaterialTheme.colorScheme.primary,
                            strokeWidth = 4.dp,
                            modifier = Modifier.size(48.dp)
                        )
                        Spacer(modifier = Modifier.height(24.dp))
                        Text(
                            text = "Memindai Musik",
                            style = MaterialTheme.typography.titleMedium,
                            fontWeight = FontWeight.Bold,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                        Spacer(modifier = Modifier.height(8.dp))
                        Text(
                            text = scanProgressText,
                            style = MaterialTheme.typography.bodyMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.7f),
                            textAlign = androidx.compose.ui.text.style.TextAlign.Center
                        )
                    }
                }
            }
        }
    }
}
